ug: Add doc on a custom CI for Apps (#457)
foundriesio/docs: source/user-guide/custom-ci/custom-ci.rst

FoundriesFactory includes all you need to build a containerized application and securely deploy it on devices.
In particular, it provides you with a git repository and the CI service that does all necessary steps
to build Apps out of source code and deliver them to devices by leveraging the TUF compliant OTA service.
You can learn more details about it by going through this :ref:`tutorial <tutorial-compose-app>`.

The FoundriesFactory solution consists of a few loosely coupled services.
It allows using the FoundriesFactory OTA framework directly eliminating a need in hosting
your App source code in the FoundriesFactory git repository and using the FoundriesFactory CI service.
Therefore, you can host your App in any source code repository and build App by any other frameworks,
yet still leverage the rest part of FoundriesFactory.

This section guides you through the steps of creating a custom CI pipeline in GitHub that:

- builds multi-arch container images and pushes them to `Foundries Registry`_;
- builds `Foundries Compose App`_ and pushes it `Foundries Registry`_;
- composes `TUF Targets role metadata`_ compliant with FoundriesFactory TUF requirements;

> - adds the composed TUF Targets to `FoundriesFactory Targets`_.

Author: fixed. What's the app, btw? Is it grammarly?

Prerequisites
-------------

> 1. Your Factory has been successfully created.
> 
> 2. At least one successful Factory CI build and corresponding Target with the tag and a hardware ID that you will be using in the following guide.
> 
> 3. You have a GitHub repo with source code, Dockerfiles and a Docker compose file that work.
> 
> For example, the prerequisites for the following guide are:
> 
> 1. Factory ``lmp-demo`` is setup and has been successfully created.
> 2. The successfully built Target with the tag ``custom-ci-devel`` and the hardware ID ``raspberrypi4-64``.

Reviewer: ```suggestion
1. Your Factory has been successfully created.

2. At least one successful Factory CI build and the corresponding Target with the tag and a hardware ID you will use in the following guide.

3. You have a GitHub repo with source code, Dockerfiles, and a Docker compose file that work.

For example, the prerequisites for the following guide are:

1. Factory ``lmp-demo`` is setup and has been successfully created.
2. The successfully built Target with the tag ``custom-ci-devel`` and the hardware ID ``raspberrypi4-64``.
```
suggestion automatically generated by the app I use to review the text

Author: fixed

            OSTree Hash:   fe15cf8ad5e09136725ef996c93299d70fa0d20bfa2f10651437b8860b9edcdb

3. `The GitHub repo`_ that contains working App implementation.

Create And Set Access Token
____________________________

The GitHub action needs to authenticate itself at the FoundriesFactory OTA service and the `Foundries Registry`_.
Therefore, an access token must be created and added to the GitHub repo action tokens prior to running any CI pipelines.
You can create the token at `FoundriesFactory WebApp`_.
The token must have `containers:read-update` and `targets:read-update` scopes to access the registry and the OTA service correspondingly.

Set Token In GitHub Repo
________________________
Go to the ``https://github.com/foundriesio/<your repo>/settings/secrets/actions`` and add a secret named `FIO_TOKEN`
and value of the token obtained in the previous step.

Define GitHub Actions Workflow
______________________________
The next step is to define a GitHub actions workflow in your repo or extend an existing one.
`The sample GitHub actions workflow`_ provides an example of the workflow that communicates with FoundiresFactory to achieve the goal,

> i.e. the items listed at the end of the first paragraph.

Author: No I meant 

```
This section guides you through the steps of creating a custom CI pipeline in GitHub that:

- builds multi-arch container images and pushes them to `Foundries Registry`_;
- builds `Foundries Compose App`_ and pushes it `Foundries Registry`_;
- composes `TUF Targets role metadata`_ compliant with FoundriesFactory TUF requirements;
- adds the composed TUF Targets to `FoundriesFactory Targets`_.
```

I suppose this is "the end of the introduction section"?

The container images hosted in the same repo as the given App and built by the repo CI job should be referred
in the compose file without any tag or hash, e.g. https://github.com/foundriesio/custom-ci-app/blob/857d78f56795c08e5f1f7bc3333f734e51468ffb/compose/home-gateway/docker-compose.yml#L13.
Container images that are not hosted in the given repo (further `external`) and are not built by the given CI job must be referred with a tag or a hash, e.g. https://github.com/foundriesio/custom-ci-app/blob/857d78f56795c08e5f1f7bc3333f734e51468ffb/compose/home-gateway/docker-compose.yml#L22.

Foundries Utilities Usage Details
_________________________________
The `compose-publish`_ CLI utility does the following (see its invocation here https://github.com/foundriesio/custom-ci-app/blob/857d78f56795c08e5f1f7bc3333f734e51468ffb/.github/workflows/fio-app-ci.yml#L89):

1. Pins images referred from the App compose file.
    a) If an image is already pinned (digest, a reference with sha256 hash) it does nothing.
    b) If an image is referred by a tag it tries to get the image digest - a reference with sha256 hash. If it fails to obtain an image digest then the utility exists with an error.

>     c) If an image reference has not tag nor hash it checks if it's specified via ``--pinned-images`` input parameter. If no digest reference found in ``pinned-images`` the the utility exists with an error.

Reviewer: ```suggestion
    c) If an image reference has not tag nor hash it checks if it's specified via ``--pinned-images`` input parameter. If no digest reference is found in ``pinned-images`` the utility exists with an error.
```

Author: fixed

2. Creates the compose App container image.
    a) Creates an archive (``tgz``) that contains the App compose file and its supplementary files.
    b) Creates a container image manifest referring to the App archive as an image layer/blob.

3. Pushes the App container image to the `Foundries Registry`_.

The utility outputs the built and pushed App image digest to the file specified via ``-d`` input parameter.
Then the published App can be referenced with a hashed URI - ``hub.foundries.io/<factory>/<app-name>@sha256:<hash>``.

> Once, the App in successfully built and pushed to the registry, a new Target referring to it can be created.

Reviewer: ```suggestion
Once the App is successfully built and pushed to the registry, a new Target referring to it can be created.
```

Author: fixed
